src/echoclient/EchoClient.php

```diff
@@ -429,6 +423,58 @@ protected function getHeaders($noCache = false)
         return $headers;
     }
 
+    /**
+     * Sends json encoded events data to echo
+     *
+     * @param string $eventsData The json encoded events data to send
+     * @return bool True if successful
+     * @throws EchoHttpException If the server responded with an error
+     * @throws EchoCouldNotSendException If we were unable to send data to the Echo server
+     */
+    protected function sendJsonEventDataToEcho($eventsData)
+    {
+        $baseUrl = $this->getBaseUrl();
+
+        if (!$baseUrl) {
+            // fail silently when creating events, should not stop user interaction as echo events are collected on a best-endeavours basis
+            $this->getLogger()->warning('Echo server is not defined (missing ECHO_HOST define), not sending events');
+            return false;
+        }
+
+        $eventUrl = $baseUrl . '/events';
+
+        try {
+            $client = $this->getHttpClient();
+            $request = $client->post($eventUrl, $this->getHeaders(), $eventsData, ['connect_timeout' => 2]);
+            $response = $request->send();
+
+            if ($response->isSuccessful()) {
+                $this->getLogger()->debug('Success sending events to echo');
+                return true;
+            } else {
+                $this->getLogger()->warning('Failed sending events to echo', [
+                    'responseCode' => $response->getStatusCode(),
+                    'responseBody' => $response->getBody(true),
+                    'batchSize' => count(json_decode($eventsData)),
+                    'batchSizeBytes' => $this->getStringSizeInBytes($eventsData),
+                ]);
+                throw new \echoclient\EchoHttpException($response->getStatusCode() . ' - ' . $response->getBody(true));
```

Comment:
do we need the response body in the exception and the log line above?

Reply:
Why not?  in the case of the exception the response body with the code provides the detail of what happened on the server. I don't see any harm in logging it here. The calling code will deal with the exception however it sees fit, but the client itself logs the details of what went wrong.